Upgrade to time 0.3

paolobarbolini opened the pull request:
Continuation of #1444

Still left to do:

* [x] Fix a solution for https://github.com/time-rs/time/discussions/356
* [x] Update documentation

Closes #1277

paolobarbolini commented:
It doesn't look like macros like it when time is renamed https://github.com/time-rs/time/blob/d700977ce9bc34013ee9841e122edbc807f410e4/time-macros/src/datetime.rs#L47

abonander commented:
Can we get around using the macros?

kyle-mccarthy commented:
I am not familiar with this issue and am not sure the circumstances are the same, but is the method used by the `postgres-types` crate possible? [It looks like they also rename the time crate](https://github.com/sfackler/rust-postgres/blob/84fa5fa1d007e95e18bef95f01e9a065ccf1e415/postgres-types/Cargo.toml#L44).

paolobarbolini commented:
> @paolobarbolini sounds like a solution wasn't forthcoming there. I dunno how feasible it is to try to get `time` to add a crate parameter to its macros, sounds like it might get stuck in bikeshedding.
> 
> Can we just inline the output of the macros here? It seems like it should be relatively stable as the code that `format_description!()` outputs doesn't appear to use any hidden types or fields.

I suggested it but they don't seem too convinced about it https://github.com/time-rs/time/discussions/356#discussioncomment-1451826

I'll try inlining the macro output.

> I am not familiar with this issue and am not sure the circumstances are the same, but is the method used by the `postgres-types` crate possible? [It looks like they also rename the time crate](https://github.com/sfackler/rust-postgres/blob/84fa5fa1d007e95e18bef95f01e9a065ccf1e415/postgres-types/Cargo.toml#L44).

That's what we're doing too

kyle-mccarthy commented:
@paolobarbolini sorry, I didn't link to the correct spot. I meant with how they [declare time as an extern crate](https://github.com/sfackler/rust-postgres/blob/84fa5fa1d007e95e18bef95f01e9a065ccf1e415/postgres-types/src/lib.rs#L219).

Doing something similar, I was able to get run `cargo check --manifest-path sqlx-core/Cargo.toml --no-default-features --features offline,all-databases,all-types,migrate,runtime-async-std-native-tls` without errors. 

The specific changes I made can be viewed here: https://github.com/kyle-mccarthy/sqlx/commit/005d9229b01918b8773fdddd49cd473fbb6afa0f

paolobarbolini commented:
> @paolobarbolini sorry, I didn't link to the correct spot. I meant with how they [declare time as an extern crate](https://github.com/sfackler/rust-postgres/blob/84fa5fa1d007e95e18bef95f01e9a065ccf1e415/postgres-types/src/lib.rs#L219).
> 
> Doing something similar, I was able to run `cargo check --manifest-path sqlx-core/Cargo.toml --no-default-features --features offline,all-databases,all-types,migrate,runtime-async-std-native-tls` without errors.
> 
> The specific changes I made can be viewed here: [kyle-mccarthy@005d922](https://github.com/kyle-mccarthy/sqlx/commit/005d9229b01918b8773fdddd49cd473fbb6afa0f)

Interesting, I didn't know extern crate had the power to rename a crate like that.

paolobarbolini commented:
I cherry-picked your commit @kyle-mccarthy. I got a warning that the `#[macro_use]` wasn't necessary, so I tried removing it.

abonander commented on sqlx-core/src/lib.rs:
```diff
@@ -104,3 +104,6 @@ pub mod mssql;
 //type HashMap<K, V> = std::collections::HashMap<K, V, ahash::RandomState>;
+
+#[cfg(feature = "time-03")]
+extern crate time_03 as time;
```
Can we make this conditional on `feature = "postgres"` as well? It should fix those "unused" warnings.

abonander replied:
Or nevermind this if you're still working on support for MSSQL/SQLite

paolobarbolini replied:
Honestly I just tested postgres. I haven't looked into mysql and sqlite, I just fixed the compiler warnings :smile:. From a quick search it looks like sqlite never supported `time`? https://github.com/launchbadge/sqlx/tree/master/sqlx-core/src/sqlite/types

abonander replied:
It definitely looks to be that way, I guess no one's ever asked for it. I guess don't worry about SQLite then?

So yeah, whatever suppresses the warning works for me.

abonander commented:
I am absolutely ecstatic that this just works! This is a backwards-compatible change too, isn't it? So we could release this with 0.5.10 instead of 0.6.0

abonander commented:
@paolobarbolini do you think you'll be able to finish this soon?

paolobarbolini commented:
> @paolobarbolini do you think you'll be able to finish this soon?

Yes, I'll look into it next week.

abonander reviewed (changes requested):
Couple remaining nits:

* can we fix the warnings on the unused `extern crate time_03 as time;` in `sqlx-core/src/lib.rs`?
* it looks like the existing type tests for `time` were repurposed to test for `time_03` which leaves `time_02` with no type tests, can we bring those back?

paolobarbolini commented:
> it looks like the existing type tests for `time 0.2` were repurposed to test the types from `time 0.3` which leaves the former with no type tests, can we make sure to cover both?

Ah right. I'll look into it

paolobarbolini commented:
I'm trying to do point 2 but we're again running into the same problem we had before with both time 0.2 and time 0.3 macros importing time as `::time`, which doesn't work because we can have only one version imported with the name `time`.

paolobarbolini commented:
It seems to me like these macro issues are going to come up again when time 0.4 comes out. Just to get an idea I looked at the downloads for the latest 0.2 [^1] and 0.3 [^2] versions and between the two time 0.3 only makes ~30% of the downloads :disappointed:. I kinda wished everybody updated so we didn't have to do this.

[^1]: https://crates.io/crates/time/0.2.27
[^2]: https://crates.io/crates/time/0.3.5

paolobarbolini commented:
Also I just realized the sqlx-macros optional feature suggestion system should be updated to support both versions :sweat_smile:

abonander commented:
@paolobarbolini should we just bite the bullet and upgrade to `time 0.3` wholesale in 0.6? Might push people to upgrade.

paolobarbolini commented:
> @paolobarbolini should we just bite the bullet and upgrade to `time 0.3` wholesale in 0.6? Might push people to upgrade.

:+1: I removed all of my changes and rebased the original author's changes on top of master

tyhi commented:
It might be worth to update the `time` version from 0.3.2 to 0.3.5, that would change MSRV 1.48 -> 1.51.

abonander commented:
Officially, we only support the latest stable version of Rust. I try to avoid using language or library features that are _too_ new but MSRV just isn't something we keep track of on this project. Should probably add a disclaimer somewhere visible for that.

thomaseizinger commented:
> It might be worth to update the `time` version from 0.3.2 to 0.3.5, that would change MSRV 1.48 -> 1.51.

This is a lib (hence no `Cargo.lock`) so unless we pin it with `=0.3.2` in the manifest, the upgrade from `0.3.2` to `0.3.5` will transparently happen in the consumer binary.

happysalada commented:
very interested in this!
It would make the move from chrono to time, much simpler.
Another potential addition, would it make sense to add the time feature serde-human-readable by default?
Without this, all the time structs are not serializable and deserializable
here is the reference https://github.com/time-rs/time/blob/main/Cargo.toml#L36
Or perhaps this should be put behind the crate's serde feature flag

paolobarbolini commented:
> Another potential addition, would it make sense to add the time feature serde-human-readable by default? Without this, all the time structs are not serializable and deserializable here is the reference https://github.com/time-rs/time/blob/main/Cargo.toml#L36 Or perhaps this should be put behind the crate's serde feature flag

The time crate docs say _Libraries should never enable this feature, as the decision of what format to use should be up to the user_

https://docs.rs/time/0.3.7/time/index.html#feature-flags

abonander commented:
Even if we wanted to enable the feature by default, which we don't for the reason @paolobarbolini stated, I have to question `time`'s choice for their human-readable serialization format, [which appears to be](https://docs.rs/time/latest/src/time/serde/mod.rs.html#92):

```
<year>-<month>-<day> <hours>:<minutes>:<seconds>.<seconds-fraction> <offset>
```

Which is _ever so close_ to being ISO-8601/RFC 3339 compliant but I believe the space separator for the offset breaks it.

abonander commented:
Indeed, it produces dates like

```
2022-02-16 20:30:08.33839334 +00:00:00
```

which are rejected by Javascript's `Date.parse()`, something we use a lot on the frontend. It's perhaps accepted by more permissive parsers like https://dencode.com/en/date/iso8601 which interpreted the above date just fine somehow, but not stricter ones. It's similarly rejected by Dart's `DateTime.parse()`, which we use in our Flutter-based mobile app projects.

You _can_ enable RFC-3339 serialization, but you have to mark every `time::OffsetDateTime` field with `#[serde(with = "time::serde::rfc3339")]`, which uses the `serde-well-known` feature, not the `serde-human-readable` feature.

The reasoning for the choice is explained here: https://github.com/time-rs/time/issues/248#issuecomment-792570479

I don't see why RFC 3339 couldn't have been the default since it's generally more useful for most of the places you'd actually use `time` with the custom extended-range format being enabled via `#[serde(with)]`, but it's a closed matter anyway.

happysalada commented:
Thanks for the comments and the correct way to serialize with serde (it all makes sense now).
Looking forward to 0.6!

kevinbader commented on sqlx-core/Cargo.toml:
```diff
@@ -151,7 +151,7 @@ sha-1 = { version = "0.9.8", default-features = false, optional = true }
 sqlformat = "0.1.8"
 thiserror = "1.0.30"
-time = { version = "0.2.27", optional = true }
+time = { version = "0.3.2", features = ["macros", "formatting", "parsing"], optional = true }
```
Wouldn't `#[serde(with = "time::serde::rfc3339")]` (as described in [this comment](https://github.com/launchbadge/sqlx/pull/1455#issuecomment-1042296608)) require the [`serde-well-known` feature flag](https://docs.rs/time/0.3.9/time/serde/index.html)? Since serializing with RFC3339 format is a very common use case, I guess adding `serde-well-known` would make sense...

abonander replied:
Generally you don't want to enable feature flags you don't strictly need. That would force that feature on for all crates that depend on SQLx.

kevinbader replied:
Makes sense... but what is then the recommended way to read a timestamp from the database and put it into a DTO in RFC3339 format? Since the version of `time` that sqlx comes with doesn't support this (due to version and that feature flag), a user of sqlx has to use an intermediate "DatabaseDTO" with `sqlx::types::time::OffsetDateTime` for fetching from the database (using `query_as`), just to then convert it manually into an "ActualDTO" with `time::OffsetDateTime` (I did that with `time::OffsetDateTime::from_unix_timestamp(db_dto.created_at.unix_timestamp())`, but I'm not sure that's the best way). Only then the aforementioned way to serialize a timestamp using RFC3339 with `with` actually works. Or am I missing something?

abonander replied:
That's only an issue because SQLx is currently using a different version of `time` than you are. Otherwise `sqlx::types::time::OffsetDateTime` and `time::OffsetDateTime` should always refer to the same type.

paolobarbolini replied:
> `time::serde::rfc3339` exists since `time` v0.3.6, so it's not available in the version proposed in this PR

sqlx depending on `0.3.2` means versions `>=0.3.2, <0.4.0` are allowed, so you're free to use features from newer patch versions of the time crate

abonander replied:
The version specified here is not restrictive and Cargo will automatically pull in the latest `0.3.x` release if it's not in your `Cargo.lock` already when building your crate. 

You should also specify `time = { version =  "0.3.6", features = ["serde-well-known"] }` in your own dependencies to be certain of this.
